Add support for http timeout config

ncjones opened the pull request:
I recently ran into an issue where CRM integrations were not succeeding but not logging failures either. It looks like it was because CRM requests were taking much longer than usual so the client was just waiting. In general client applications should always have a finite timeout set so that badly behaving services like this do not cause hard-to-diagnose cascading failure due to resource exhaustion (in my case this scenario is likely to cause pubsub consumers to be exchausted).

Let me know what you think of this change. I still need to update the XHR client to be compatible with the new method signature. I think in general if a JS method has more than two args then an options parameter should be used. The syntax I've used here is the old verbose style because I'm not sure what the target runtime is. It could be improved by using argument destructuring if that's an option.

AleksandrRogov commented:
Hi @ncjones ,
Everything looks good to me. `options` is good as well. I did not have the `timeout` issue yet, but it can happen at any time, thanks for pointing this out.
ES5 is good enough for me.
Can you please include a change for `xhr` to a pull request as well?
Much appreciated

ncjones commented:
@AleksandrRogov I've updated the PR to make the XHR transport compatible with new transport options signature but I have not implemented timeout for XHR as I'm not set up for testing that. I renamed "async" to "isAsync" to avoid using an ES6 keyword.

AleksandrRogov commented:
@ncjones for some reason `request.on('timeout', ...)` not working for me, I had to change it to `request.setTimeout(options.timeout, ...)`, so the timeout callback is called. Did you have this issue?

The code is the following now:
```js
if (options.timeout) {
  request.setTimeout(options.timeout, function () {
     request.abort();
  });
}
````

Also, when the request aborted the error message becomes the following: "Request timed out: Error: socket hang up". Can I remove this part?
```js
if (request.aborted) {
  error = new Error('Request timed out: ' + error);
}
```
